cli/src/main.rs

```diff
@@ -60,7 +62,11 @@ async fn main() -> color_eyre::Result<()> {
             allow_existing,
         } => {
             let user_id = get_user_id_from_session()?;
-            append_url_to_db(&db_pool, &url, user_id, allow_existing).await?;
+            append_url(&db_pool, &url, user_id, allow_existing).await?;
+            // let outcome = append_url(&db_pool, &url, user_id, allow_existing).await?;
+            // if let AddUrlOutcome::Created(page) | AddUrlOutcome::Existing(page) = outcome {
```

Comment:
Try something for me but it might not work 😆 

I want to see if this works without the `if` so just starting with `let AddUrlOutcome::Created(page) | AddUrlOutcome::Existing(page)`

The reason is a bit subtle lol! What I _want_ to happen is for Rust to error out if we ever added a third type to this enum

What you have now says "If the outcome is a Created or Existing, take its page out and do stuff with it, but if it does NOT match we simply ignore it"

This works great today, cause the enum HAS to be Created or Existing. But what if we added a third variant? In the `if let` version Rust would happily compile and just skip this code path for the third variant.

But I don't think thats _really_ what we want here! We'd RATHER if the Rust compiler errored out when we add the third variant, and tell us we need to update this code. I believe that is what the `let` alone will do! It should work now when we handle both the variants, but if a third is added that we don't handle I believe it will start erroring out!

Does that make sense? And does this code work with just the `let`? lol

Reply:
Okay I'm not sure if I did this entirely correct but I think this maybe could solve the problem?

`            
let page = match outcome {
                AddUrlOutcome::Created(page) | AddUrlOutcome::Existing(page) => page,
};

process_page_snapshot(&db_pool, page).await?;
`

Reply:
Yes this is PERFECT!! Leaning into the Rust pattern matching is awesome!

Here are a few different versions of this just to compare! There are all just 'syntactic sugar' meaning they will literally be compiled to the same things at the end of the day, just whichever you prefer the style of!

Here is what you wrote with my syntax highlighting just to have it as a comparison!

<img width="785" alt="image" src="https://github.com/coreyja/knowledge/assets/964771/8139ffa6-142b-476b-ba8b-8a6b46c5d7e1">

Here is the first change! Since we only have a single branch on our `match` we can actually just remove it and do the pattern matching in the `let`!

<img width="761" alt="image" src="https://github.com/coreyja/knowledge/assets/964771/aa717b4e-efa1-48db-a136-660ee9433fc0">

And if we wanted to push that one farther we could remove the `outcome` var and do this all on one line!

<img width="693" alt="image" src="https://github.com/coreyja/knowledge/assets/964771/5e592896-2386-49ae-8884-33ec6a22f55a">

And that last one is _basically_ what you had before without the leading `if` I believe!

One way to think about the `if` at the beginning is `if this pattern matches`, where if the pattern does NOT match it will just just skip that block.
Removing the `if` changes that to a compile error if we know we can't match the pattern!